File: src/test/java/bad/robot/radiate/AggregatorTest.java
package bad.robot.radiate;

import bad.robot.radiate.teamcity.Any;
import com.googlecode.totallylazy.Sequences;
import org.junit.Test;

import static bad.robot.radiate.Aggregator.aggregate;
import static com.googlecode.totallylazy.Sequences.sequence;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

public class AggregatorTest {

    @Test
    public void aggregateProgress() {
        Aggregator aggregate = aggregate(Sequences.sequence(
            Any.runningBuildPercentageCompleteAt(20),
            Any.build(),
            Any.runningBuildPercentageCompleteAt(2),
            Any.runningBuildPercentageCompleteAt(3)));
        assertThat(aggregate.progress().toString(), is("8%"));
    }

    @Test
    public void aggregateNumberBuilds() {
        Aggregator aggregate = aggregate(sequence(
                Any.runningBuildPercentageCompleteAt(20),
                Any.build(),
                Any.runningBuildPercentageCompleteAt(2),
                Any.runningBuildPercentageCompleteAt(3)));
        Progress progress = aggregate.progress();
        assertThat(progress.numberOfBuilds(), is(3));
    }

    @Test
    public void nullObject() {
        Aggregator aggregate = aggregate(sequence(Any.build(), Any.runningBuildPercentageCompleteAt(2)));
        assertThat(aggregate.progress().numberOfBuilds(), is(1));
    }

    @Test
    public void noProgress() {
        Aggregator aggregate = aggregate(sequence(Any.build(), Any.build(), Any.build()));
        assertThat(aggregate.progress().numberOfBuilds(), is(0));
        assertThat(aggregate.progress().toString(), is("0%"));
    }
}

File: src/test/scala/bad/robot/radiate/AggregatorTest.scala
package bad.robot.radiate

import bad.robot.radiate.teamcity.{Any => Any}
import org.specs2.mutable.Specification

class AggregatorTest extends Specification {

  "Aggregate progress" >> {
    val aggregate = new Aggregator(List(
      Any.runningBuildPercentageCompleteAt(20),
      Any.build,
      Any.runningBuildPercentageCompleteAt(2),
      Any.runningBuildPercentageCompleteAt(3)))
    aggregate.progress.toString must_== "8%"
  }

  "Aggregate number of builds" >> {
    val aggregate = new Aggregator(List(
      Any.runningBuildPercentageCompleteAt(20),
      Any.build,
      Any.runningBuildPercentageCompleteAt(2),
      Any.runningBuildPercentageCompleteAt(3)))
    aggregate.progress.numberOfBuilds must_== 3
  }

  "Null object" >> {
    val aggregate = new Aggregator(List(Any.build, Any.runningBuildPercentageCompleteAt(2)))
    aggregate.progress.numberOfBuilds must_== 1
  }

  "No progress" >> {
    val aggregate = new Aggregator(List(Any.build, Any.build, Any.build))
    aggregate.progress.numberOfBuilds must_== 0
    aggregate.progress.toString must_== "0%"
  }
}
